src/run-graph.ts

```diff
@@ -305,12 +307,49 @@ export class RunGraph {
     return pkgsInError.length > 0
   }
 
-  expandGlobs(globs: string[]) {
-    return this.pkgJsons.map(p => p.name).filter(name => globs.some(glob => minimatch(name, glob)))
+  expandGlobs(globs: string[]): Promise<string[]> {
+    let pkgs = this.pkgJsons
+      .map(p => p.name)
+      .filter(name => globs.some(glob => minimatch(name, glob)))
+
+    // if changedSince is defined, filter the packages to contain only changed packages (according to git)
+    if (this.opts.changedSince) {
+      return getChangedFilesForRoots([this.opts.workspacePath], {
+        changedSince: this.opts.changedSince
+      }).then(data => {
+        if (!data.repos || (data.repos.git.size === 0 && data.repos.hg.size === 0)) {
+          throw new Error(
+            "The workspace is not a git/hg repo and it cannot work with 'changedSince'"
+          )
+        }
+
+        /**
+         * filter the packages by checking if they have any changed files. This way is quicker
+         * (mapping over packages) because package count is usually lower than changed files count
+         * and we only need to check once per package.
+         */
+        pkgs = pkgs.filter(pkg => {
```

Comment:
we can extract this into a standalone function and add some .spec.ts for it

Reply:
Done. Resolving this.